shell/shared/shell.js

```diff
@@ -330,48 +379,70 @@ if (Meteor.isClient) {
       if (actionId === "dev") {
         var devId = event.currentTarget.getAttribute("data-devid");
         var devIndex = event.currentTarget.getAttribute("data-index");
-        var devApp = DevApps.findOne(devId);
-        if (!devApp) {
-          console.error("no such dev app: ", devId);
-          return;
-        }
+      }
 
-        var devAction = devApp.manifest.actions[devIndex];
+      launchAndEnterGrainByActionId(actionId, devId, devIndex);
+    },
 
-        packageId = devApp.packageId;
-        command = devAction.command;
-        actionTitle = devAction.title.defaultText;
-      } else {
-        var action = UserActions.findOne(actionId);
-        if (!action) {
-          console.error("no such action: ", actionId);
-          return;
-        }
+    "click .action-required button": function (event) {
+      event.currentTarget.parentNode.parentNode.style.display = "none";
+    },
+  });
 
-        packageId = action.packageId;
-        command = action.command;
-        actionTitle = action.title;
-      }
+  Template.appdemo.helpers({
+    appName: function() {
+      // we ask the server for this in appdemo.js
+      return Session.get('appName');
+    }
+  });
+
+  Template.appdemo.events({
+    "click #createDemoUser": function (event) {
+      /*
+	When clicking on the createDemoUser button on the app demo,
+	we want to:
+
+	1. Create the Demo User.
+
+	2. Log the user in as this Demo User.
+
+	3. Install the chosen app.
+
+	4. Create a new grain with this app.
 
-      var title = actionTitle;
-      if (title.lastIndexOf("New ", 0) === 0) {
-        title = actionTitle.slice(4);
+	5. Take them into this grain.
+
+      */
+      // calculate the appId
+
+      // 1. Create the Demo User & 2. Log the user in as this Demo User.
+      var displayName = document.getElementById("demo-display-name").value.trim();
+      if (displayName === "") {
+        displayName = "Demo User";
+      } else {
+        displayName += " (demo)";
       }
-      title = "Untitled " + title;
-
-      // We need to ask the server to start a new grain, then browse to it.
-      Meteor.call("newGrain", packageId, command, title, function (error, grainId) {
-        if (error) {
-          console.error(error);
-        } else {
-          Router.go("grain", {grainId: grainId});
+
+      Accounts.callLoginMethod({
+        methodName: "createDemoUser",
+        methodArguments: [displayName],
+        userCallback: function (err) {
```

Comment:
A somewhat more meteor-magic way to do this would be to have /appdemo/:packageId go directly to installing and opening the app if the user is already logged in. Then you don't need a login callback here, because once the login is complete the page will automatically re-render because Meteor.userId() is a reactive data source. :)